src/core/image_utils.js

```diff
@@ -0,0 +1,184 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/* eslint no-var: error */
+
+import { assert, OPS, warn } from '../shared/util';
+import { ColorSpace } from './colorspace';
+import { JpegStream } from './jpeg_stream';
+import { Stream } from './stream';
+
+class NativeImageDecoder {
+  constructor({ xref, resources, handler, forceDataSchema = false,
+                pdfFunctionFactory, }) {
+    this.xref = xref;
+    this.resources = resources;
+    this.handler = handler;
+    this.forceDataSchema = forceDataSchema;
+    this.pdfFunctionFactory = pdfFunctionFactory;
+  }
+
+  canDecode(image) {
+    return image instanceof JpegStream &&
+           NativeImageDecoder.isDecodable(image, this.xref, this.resources,
+                                          this.pdfFunctionFactory);
+  }
+
+  decode(image) {
+    // For natively supported JPEGs send them to the main thread for decoding.
+    const dict = image.dict;
+    let colorSpace = dict.get('ColorSpace', 'CS');
+    colorSpace = ColorSpace.parse(colorSpace, this.xref, this.resources,
+                                  this.pdfFunctionFactory);
+
+    return this.handler.sendWithPromise('JpegDecode', [
+      image.getIR(this.forceDataSchema), colorSpace.numComps
+    ]).then(function({ data, width, height, }) {
+      return new Stream(data, 0, data.length, dict);
+    });
+  }
+
+  /**
+   * Checks if the image can be decoded and displayed by the browser without any
+   * further processing such as color space conversions.
+   */
+  static isSupported(image, xref, res, pdfFunctionFactory) {
+    const dict = image.dict;
+    if (dict.has('DecodeParms') || dict.has('DP')) {
+      return false;
+    }
+    const cs = ColorSpace.parse(dict.get('ColorSpace', 'CS'), xref, res,
+                                pdfFunctionFactory);
+    // isDefaultDecode() of DeviceGray and DeviceRGB needs no `bpc` argument.
+    return (cs.name === 'DeviceGray' || cs.name === 'DeviceRGB') &&
+           cs.isDefaultDecode(dict.getArray('Decode', 'D'));
+  }
+
+  /**
+   * Checks if the image can be decoded by the browser.
+   */
+  static isDecodable(image, xref, res, pdfFunctionFactory) {
+    const dict = image.dict;
+    if (dict.has('DecodeParms') || dict.has('DP')) {
+      return false;
+    }
+    const cs = ColorSpace.parse(dict.get('ColorSpace', 'CS'), xref, res,
+                                pdfFunctionFactory);
+    const bpc = dict.get('BitsPerComponent', 'BPC') || 1;
+    return (cs.numComps === 1 || cs.numComps === 3) &&
+           cs.isDefaultDecode(dict.getArray('Decode', 'D'), bpc);
+  }
+}
+
+const ImageCacheKind = {
+  NON_EXISTING: 0,
+  EXISTING: 1,
+  NEEDS_RESIZING: 2,
+};
+
+class ImageCache {
+  constructor({ handler, pageIndex, transform = null, }) {
+    this._handler = handler;
+    this._pageIndex = pageIndex;
+    this._transform = transform;
+    this._cache = Object.create(null);
+  }
+
+  get({ key, }) {
```

Comment:
If we only have a key, we don't need a parameter object here I think, unless you want to extend this later on?

Reply:
This may indeed be needed later, and I don't want to have to (unnecessarily) update the call-sites.